File: 11-container-with-most-water/11-container-with-most-water.js
/**
 * @param {number[]} height
 * @return {number}
 */
var maxArea = function(height) {
    let left = 0
    let right = height.length -1
    let result = 0
    while (left < right) {
      const l = height[left]
      const r = height[right]
      if (l < r) {
        result = Math.max(result, (right-left)*l)
        left++
      } else if (r < l) {
        result = Math.max(result, (right-left)*r)
        right--
      } else {
        result = Math.max(result, (right-left)*l)
        left++
        right--
      }
    }
    return result
};

File: 11-container-with-most-water/11-container-with-most-water.scala
object Solution {
  def maxArea(height: Array[Int]): Int = {
    var li = 0
    var ri = height.length - 1
    var l = height(li)
    var r = height(ri)
    var maxH = 0
    var max = 0
    while (li < ri) {
      val minH = math.min(l, r)
      if (minH > maxH) {
        maxH = minH
        max = math.max(max, maxH * (ri - li))
      }
      if (l < r) {
        li += 1
        l = height(li)
      } else if (l > r) {
        ri -= 1
        r = height(ri)
      } else {
        li += 1
        l = height(li)
        ri -= 1
        r = height(ri)
      }
    }
    max
  }
}